pageserver/src/walredo.rs

```diff
@@ -332,9 +334,17 @@ impl PostgresRedoManager {
             anyhow::bail!("invalid neon WAL redo request with no base image");
         }
 
-        // Apply all the WAL records in the batch
-        for (record_lsn, record) in records.iter() {
-            self.apply_record_neon(key, &mut page, *record_lsn, record)?;
+        // process the records in batches and yield; this should guard against pathological
+        // situations where we accidentially have a huge number of in-neon applied records.
+        let yield_every = 200;
+
+        for records in records.chunks(yield_every) {
```

Comment:
I'd prefer just one loop with `records.into_iter().zip((0..yield_every).rev().cycle())` with a
`if v == 0 {yield_now().await;}` or something like that.

Reply:
Please elaborate, I do like combinators but I do think this solution is more readable.